### PR TITLE
JS: lean dependency installation

### PR DESCRIPTION
**Pull request by asgerf**
node_modules_size

# Lean dependency installation
This replaces the `yarn`-based dependency installation with a "lean" dependency installer which only installs what we need for TypeScript type checking, saving time and space. Dependency installation is still behind a feature flag, but hopefully we can soon turn it on by default now that it's faster.

Main advantages are:
- Much faster, see below
- Deterministic* dependency version resolution.
- We do not require `yarn` to be installed
- Secure by default. No quirks like `.yarnrc` that we have to opt out of to maintain extraction safety. We don't chmod extracted files, we only extract files with a known extension etc.

### Deterministic version resolution

NPM and Yarn will generally try to install the latest viable version of each transitive dependency, which can change as more packages are published. Instead, we aim for the most recent version that's explicitly mentioned in a dependency constraint (excluding upper bound constraints like `< 2.0.0`).

*The only source of non-deterministism that I'm aware of is that we treat network failures as non-existing packages, and non-existing packages are simply ignored.

### Installation time comparison
Some numbers comparing the installation time and size of `node_modules` when comparing the old method to the new method, when installing the top-level `package.json` from various repositories:

Overall it's about 6-20x faster and uses 100-500x less disk space.

| repository | Yarn-time | Yarn-size | Lean-time | Lean-size |
|------------|-----------|-----------|-----------|-----------|
angular | 64.97s | 1.2G | 10.03s | 13M
| vscode     | 25.37s    | 373M      | 3.53s     | 1.6M      |
juice-shop | 47.47s | 262M | 2.32s | 396K
angular-gantt | 17.33s | 200M | 2.12s | 316K

### Evaluation (compared to no dependency installation at all)

- [typescript-full](https://git.semmle.com/asger/dist-compare-reports/tree/js/lean-dependency-installation-java-cleanhist_1600353168863) shows a large number of new typed expressions and a couple of new call edges.
- I'm running more evaluations and will measure final extraction time.

### Other stuff

There's also a commit making the TypeScript extractor aware of the `types/typings` field of `package.json` files, which is needed to properly associate module names with symbols. This is needed to make things like `hasUnderlyingType("@angular/common", "blah")` work correctly.

This depends on an internal PR that adds `common-compress.jar` as a dependency (for unpacking tar archives).

Commit-by-commit review strongly recommended. For the third commit, I suggest reviewing files in this order:
- `Fetcher`, `AsyncFetcher`, `DependencyResolver`, `AutoBuild`

### CODE REVIEW

File: javascript/extractor/src/com/semmle/js/dependencies/VersionConstraint.java
Comment (erik-krogh) [inline javascript/extractor/src/com/semmle/js/dependencies/VersionConstraint.java:32]: I don't get this line. 
[The documentation](https://docs.npmjs.com/files/package.json#dependencies) suggests that `||` and not `,` is used for multiple constraints.

File: javascript/extractor/src/com/semmle/js/dependencies/VersionConstraint.java
Comment (erik-krogh) [inline javascript/extractor/src/com/semmle/js/dependencies/VersionConstraint.java:45]: How about `1.0.x`? 

Also: "1.x" and ">= 1.0" are not equivalent (https://semver.npmjs.com/).

**Review by erik-krogh — COMMENTED** [review]
I'm about half-way with a first pass through the code, but I think I got enough comments for now. 

Some tests of the new behavior could be nice.  
I get that fetching packages during testing is probably a bad idea.  
But I think it should be possible to have a test for the `typings`/`types` field support? 

I approved the internal PR, so hopefully the tests can pass once that is merged.

**Comment by asgerf** [comment]
Some more detailed extraction times shows that extraction time actually goes up by quite a bit due to TypeScript compiler having more work to do (while actual dependency installation is relatively inexpensive):
```
project              before  after  install  relative
tslint               35s     36s    (2s)     1.02857
TypeScript           165s    176s   (1s)     1.06667
cherow               61s     66s    (4s)     1.08197
vscode               469s    518s   (5s)     1.10448
glimmer-vm           46s     51s    (3s)     1.1087
tinymce              169s    190s   (4s)     1.12426
shumway              51s     73s    (0s)     1.43137
pluotsorbet          17s     25s    (0s)     1.47059
Stopify              23s     34s    (2s)     1.47826
kibana               1262s   1901s  (11s)    1.50634
angular              415s    711s   (6s)     1.71325
insight              11s     21s    (5s)     1.90909
ORCID-Source         25s     50s    (2s)     2
jhipster-sample-app  15s     30s    (5s)     2
angularjs-slider     10s     21s    (0s)     2.1
esprima              16s     38s    (0s)     2.375
onos                 24s     61s    (3s)     2.54167
juice-shop           23s     61s    (4s)     2.65217
angular-gantt        31s     142s   (3s)     4.58065
xenon                27s     135s   (5s)     5
```

It's still behind a feature flag and the installation-time looks good, so I don't think this blocks the PR, but we may need to implement some extractor performance improvements before turning it on by default.

File: javascript/extractor/lib/typescript/src/main.ts
Comment (asgerf) [reply javascript/extractor/lib/typescript/src/main.ts:100]: I tried using `object` at first, but it does not allow access to its members, so we'd need a type cast on this line:
```ts
let typings = json.types || json.typings;
```
We could define an interface for `PackageJson` but then we'd be using that interface without knowing for sure that the value actually satisfies it. Ultimately I think `any` is fine when dealing with unchecked JSON data.

File: javascript/extractor/src/com/semmle/js/dependencies/Fetcher.java
Comment (asgerf) [reply javascript/extractor/src/com/semmle/js/dependencies/Fetcher.java:130]: I'd prefer to keep a defensive style throughout - we often run into bugs that aren't technically our fault but end up being our problem anyway. There's only _supposed_ to be one top-level entry, but the tarballs were also supposed to gzipped and that wasn't the case.

File: javascript/extractor/src/com/semmle/js/dependencies/VersionConstraint.java
Comment (asgerf) [reply javascript/extractor/src/com/semmle/js/dependencies/VersionConstraint.java:32]: Oh you're right that's completely wrong 🤦

File: javascript/extractor/src/com/semmle/js/dependencies/VersionConstraint.java
Comment (asgerf) [reply javascript/extractor/src/com/semmle/js/dependencies/VersionConstraint.java:45]: Right, I'll make it a little more clear that we just parse the constraints in a way that works for our resolution algorithm (i.e. when picking only explicitly-mentioned versions, the range semantics aren't important). We just treat the `x` as a zero for this purpose.

**Comment by asgerf** [comment]
@erik-krogh Instead of trying to parse the constraints properly, I rewrote it to just parse out the bits we actually care about. PTAL.

**Review by erik-krogh — COMMENTED** [review]
A few more thoughts and comments. 

Do you want me to merge the internal PR so we can get the tests to go green on this PR?

File: javascript/extractor/src/com/semmle/js/dependencies/DependencyResolver.java
Comment (asgerf) [reply javascript/extractor/src/com/semmle/js/dependencies/DependencyResolver.java:114]: It gets really fiddly in the recursive case where we have to concatenate lists. It seems less error prone to have one list where we put everything so nothing can get lost.

File: javascript/extractor/src/com/semmle/js/dependencies/AsyncFetcher.java
Comment (asgerf) [reply javascript/extractor/src/com/semmle/js/dependencies/AsyncFetcher.java:44]: I believe the convention is to use `LinkedHashMap` by default.

File: javascript/extractor/src/com/semmle/js/dependencies/Fetcher.java
Comment (asgerf) [reply javascript/extractor/src/com/semmle/js/dependencies/Fetcher.java:107]: I'll add a comment to clarify why, but unfortunately not. The two streams are compressed for very different reasons and require different means for determining the compression format.

The packument is gzipped because we use the `Accept-Encoding` header, i.e. it's compressed as part of the HTTP layer. The compression format can thus be deduced from the `Content-Encoding` header.

For the tarball, the compression format is not determined by the `Accept-Encoding` header. Setting `Accept-Encoding` here has no effect and the `Content-Encoding` is never `gzip`. Think of it as a file sent via an uncompressed HTTP stream, and that file happens to be gzip-compressed in some cases.

File: javascript/extractor/src/com/semmle/js/dependencies/DependencyResolver.java
Comment (asgerf) [reply javascript/extractor/src/com/semmle/js/dependencies/DependencyResolver.java:65]: Would rather keep the current behaviour. An explicitly-mentioned version at least has a high likelihood of working, because it most likely worked at the point that constraint was added.

File: javascript/extractor/src/com/semmle/js/dependencies/AsyncFetcher.java
Comment (erik-krogh) [reply javascript/extractor/src/com/semmle/js/dependencies/AsyncFetcher.java:44]: You're right, it looks like we use `LinkedHashMap` everywhere.

File: javascript/extractor/src/com/semmle/js/dependencies/AsyncFetcher.java
Comment (asgerf) [reply javascript/extractor/src/com/semmle/js/dependencies/AsyncFetcher.java]: The other class, `CachedOperation`, can't be static as it refers to the `executor` field, but I've made it private.

**Comment by asgerf** [comment]
@erik-krogh Merging the internal PR sounds like a good idea

File: javascript/extractor/src/com/semmle/js/dependencies/DependencyResolver.java
Comment (erik-krogh) [reply javascript/extractor/src/com/semmle/js/dependencies/DependencyResolver.java:115]: Right. But you can have a single list as a local variable by using an accumulator.   
Like this: https://github.com/github/codeql/commit/9ecc106562e1a73fb152b92c07cf5cb3a63adbc2

Comment (asgerf) [reply javascript/extractor/src/com/semmle/js/dependencies/DependencyResolver.java:115]: Not really seeing the benefit here.

Comment (erik-krogh) [reply javascript/extractor/src/com/semmle/js/dependencies/DependencyResolver.java:115]: You would avoid having a field, and instead have a local variable.  
But yeah, it comes with some complexity.

**Comment by asgerf** [comment]
Rebased to resolve trivial conflict.

**Review by erik-krogh — COMMENTED** [review]
I think these two changes will make the build go green. (The `Pull-Request-Check/Java-Checks`)

**Comment by asgerf** [comment]
> Some tests of the new behavior could be nice.

I'll add some AutoBuild tests with dependency installation.

**Review by erik-krogh — COMMENTED** [review]
:+1:  

You just need to resolve the conflict.

**Review by erik-krogh — APPROVED** [review]

**Pull request merged**
